### PR TITLE
MESH-1270: Link Documentation to Corporate Actions

### PR DESCRIPTION
**Pull request by Centril**
- `DocumentName` as a key is moved towards `DocumentId`.
- Also some misc refactoring.

### CODE REVIEW

File: pallets/corporate-actions/src/lib.rs [hunk 1]
```diff
@@ -425,6 +433,40 @@ decl_module! {
...
+            Self::ensure_ca_exists(id)?;
+            for doc in &docs {
+                <Asset<T>>::ensure_doc_exists(&id.ticker, doc)?;
+            }
+
+            // TODO(Centril): Consider whether we want to dedup `docs` here.
+            // This will be considerably faster once `docs` is a vector of IDs.
+            // It would turn what is currently an `O(1)` operation in terms of
+            // local work (not storage reads!) into an `O(n log n)` operation.
+            // The benefits of deduping here are not entirely obvious.
```
Comment (Centril) [inline pallets/corporate-actions/src/lib.rs:464]: @maxsam4 please take a look at this.

Comment (maxsam4) [reply pallets/corporate-actions/src/lib.rs:464]: I wouldn't bother with it. I don't see any benefit of deduping it.

File: pallets/corporate-actions/src/lib.rs [hunk 2]
```diff
@@ -425,6 +433,34 @@ decl_module! {
...
+        /// ## Arguments
+        /// - `id` of the CA to associate with `docs`.
+        /// - `docs` to associate with the CA with `id`.
+        ///
+        /// # Errors
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `NoSuchCA` if `id` does not identify an existing CA.
+        /// - `NoSuchDoc` if any of `docs` does not identify an existing document.
+        #[weight = <T as Trait>::WeightInfo::link_ca_doc(docs.len() as u32)]
+        pub fn link_ca_doc(origin, id: CAId, docs: Vec<DocumentId>) {
```
Comment (maxsam4) [inline pallets/corporate-actions/src/lib.rs:452]: Perhaps add a wrapper function that takes in a Document, stores it in the asset module, and links it in this module.

File: pallets/corporate-actions/src/lib.rs [hunk 3]
```diff
@@ -425,6 +433,34 @@ decl_module! {
...
+            Self::ensure_ca_exists(id)?;
+            for doc in &docs {
+                <Asset<T>>::ensure_doc_exists(&id.ticker, doc)?;
+            }
+
+            // Add the link and emit event.
+            CADocLink::mutate(id, |slot| *slot = docs.clone());
+            Self::deposit_event(Event::CALinkedToDoc(caa, id, docs));
+        }
     }
```
Comment (maxsam4) [inline pallets/corporate-actions/src/lib.rs:464]: `initiate_corporate_action` should take in a vec<doc> and vec<doc_id> to add/link documents.

**Review by maxsam4 — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Review by satyamakgec — APPROVED** [review]

**Pull request merged**